### ISSUE TITLE
Simple ask

### ISSUE THREAD

**Issue by UserXGnu** [gwutz/gnome-university-code-examples#1]
Wutzara,
First, sorry about my english, it is too bad, I know. (I'm from Brazil, so ...)
I am very excited to learn the apis GTK and Gnome, but there are some things I do not understand, for example:
  GtkWidget \* window = gtk_window_new (GTK_WINDOW_TOPLEVEL);

Here I think I created a GtkWindow in a GtkWidget, this was possible because of inheritance. But up front, in your example (src / basic-gtk / basic.c):

gtk_signal_connect (GTK_WIDGET (window), "destroy", G_CALLBACK (gtk_main_quit), NULL);

in the beginning you do a casting GTK_WIDGET (window), but window is a widget, or I misunderstood? I really need to do this casting? Because without casting works, but I fear that might work incorrectly.

Again, sorry about my english and thank you for your help !!!

**Comment by gwutz** [comment]
Hey,
no problem. And i'm no native speaker so my english is also not the best :). Its correct, this cast is not necessary. g_signal_connect is in gsignal.h defined with

``` c
#define g_signal_connect(instance, detailed_signal, c_handler, data) \
    g_signal_connect_data ((instance), (detailed_signal), (c_handler), (data), NULL, (GConnectFlags) 0)
gulong   g_signal_connect_data            (gpointer       instance,
                           const gchar   *detailed_signal,
                           GCallback      c_handler,
                           gpointer       data,
                           GClosureNotify     destroy_data,
                           GConnectFlags      connect_flags);
```

So a simple void pointer is also okay.

Its a good question what i thought about this casts :) If you use Methods (for example gtk_container_add(..)) its necessary to avoid gcc warnings (and in C context its important). The background about this cast is the struct interleaving.

Example:

``` c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int a;
} A;

typedef struct {
    A parent;
    int b;
} B;

int
main (int argc, char *argv[])
{
    B* test = malloc(sizeof(B));

    test->b = 10;

    printf("b=%d\n", test->b);

    // This is not possible!
    //test->a = 20;

    // This is not nice!
    //test->parent.a = 20;

    ( (A*) test)->a = 20;   
    printf("a=%d\n", ((A*)test)->a);

    return 0;
}
```

Feel free to ask me :)

**Comment by UserXGnu** [comment]
Wutzara, =-D
So fast to answer...
Thank you for your explanation. It made me understand why every code I see uses this casting ...
I program in C as well (at least I think so), but sometimes, some new things make me a little lost. Many things to learn!! 

surely there will be many new questions =-D

Thanks.

**Comment by UserXGnu** [comment]
Hey man,

I was seeing your examples of GU again and then I see: 

void
gu_textwindow_finalize(GObject _object) {
   GuTextwindowPrivate_ priv;
   priv = GU_TEXTWINDOW(object)->priv;

   G_OBJECT_CLASS(gu_textwindow_parent_class)->finalize(object);
}

so, I did not understand that small part: GuTextwindowPrivate \* priv; priv = GU_TEXTWINDOW(object)->priv, couse after doing that you do not used "priv" anyware is that right or am I saying some stupid thing?

Once more, Thank you for your help.

**Comment by gwutz** [comment]
Thats correct, i dont use the private structure in this example. But it is really annoying to add it everytime you want to use it. Most Gtk programms have this two-liner in the start of a function. But in this special case it is absolute useless. :)
